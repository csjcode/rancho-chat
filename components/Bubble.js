import * as Clipboard from 'expo-clipboard'

import { Feather, FontAwesome } from '@expo/vector-icons'
import {
  Image,
  StyleSheet,
  Text,
  TouchableWithoutFeedback,
  View,
} from 'react-native'
import {
  Menu,
  MenuOption,
  MenuOptions,
  MenuTrigger,
} from 'react-native-popup-menu'
import React, { useRef } from 'react'

import { formatAmPm } from '../utils/datetimeFormat'
import getColors from '../constants/colors/getColors'
import { starMessage } from '../utils/actions/chatActions'
import { useSelector } from 'react-redux'
import uuid from 'react-native-uuid'
const colorsTheme = getColors()

const MenuItem = (props) => {
  const colorsTheme = getColors()
  const Icon = props.iconPack ?? Feather

  return (
    <MenuOption onSelect={props.onSelect}>
      <View style={stylesFor(colorsTheme).menuItemContainer}>
        <Text style={stylesFor(colorsTheme).menuText}>{props.text}</Text>
        <Icon name={props.icon} size={18} />
      </View>
    </MenuOption>
  )
}

const Bubble = (props) => {
  const {
    text,
    type,
    messageId,
    chatId,
    userId,
    date,
    setReply,
    replyingTo,
    name,
    imageUrl,
  } = props

  const starredMessages = useSelector(
    (state) => state.messages.starredMessages[chatId] ?? {},
  )
  const storedUsers = useSelector((state) => state.users.storedUsers)

  const bubbleStyle = { ...stylesFor(colorsTheme).container }
  const textStyle = { ...stylesFor(colorsTheme).text }
  const wrapperStyle = { ...stylesFor(colorsTheme).wrapperStyle }

  const menuRef = useRef(null)
  const id = useRef(uuid.v4())

  let Container = View
  let isUserMessage = false
  const dateString = date && formatAmPm(date)

  switch (type) {
    case 'system':
      textStyle.color = '#65644A'
      bubbleStyle.backgroundColor = colorsTheme.beige
      bubbleStyle.alignItems = 'center'
      bubbleStyle.marginTop = 10
      break
    case 'error':
      bubbleStyle.backgroundColor = colorsTheme.red
      textStyle.color = 'white'
      bubbleStyle.marginTop = 10
      break
    case 'myMessage':
      wrapperStyle.justifyContent = 'flex-end'
      bubbleStyle.backgroundColor = colorsTheme.bubbleBackgroundColor
      textStyle.color = colorsTheme.bubbleTextColor
      bubbleStyle.maxWidth = '90%'
      Container = TouchableWithoutFeedback
      isUserMessage = true
      break
    case 'theirMessage':
      wrapperStyle.justifyContent = 'flex-start'
      bubbleStyle.maxWidth = '90%'
      Container = TouchableWithoutFeedback
      isUserMessage = true
      textStyle.color = colorsTheme.bubbleTextColor
      break
    case 'reply':
      bubbleStyle.backgroundColor = colorsTheme.bubbleBackgroundColorReply
      textStyle.color = colorsTheme.bubbleTextColor
      break
    case 'info':
      bubbleStyle.backgroundColor = 'white'
      bubbleStyle.alignItems = 'center'
      textStyle.color = colorsTheme.textColor
      break
    default:
      break
  }

  const copyToClipboard = async (text) => {
    try {
      await Clipboard.setStringAsync(text)
    } catch (error) {
      console.log(error)
    }
  }

  const isStarred = isUserMessage && starredMessages[messageId] !== undefined
  const replyingToUser = replyingTo && storedUsers[replyingTo.sentBy]

  return (
    <View style={wrapperStyle}>
      <Container
        onLongPress={() =>
          menuRef.current.props.ctx.menuActions.openMenu(id.current)
        }
        style={{ width: '100%' }}
      >
        <View style={bubbleStyle}>
          {name && type !== 'info' && (
            <Text style={stylesFor(colorsTheme).name}>{name}</Text>
          )}

          {replyingToUser && (
            <Bubble
              type="reply"
              text={replyingTo.text}
              name={`${replyingToUser.firstName} ${replyingToUser.lastName}`}
            />
          )}

          {!imageUrl && <Text style={textStyle}>{text}</Text>}

          {imageUrl && (
            <Image
              source={{ uri: imageUrl }}
              style={stylesFor(colorsTheme).image}
            />
          )}

          {dateString && type !== 'info' && (
            <View style={stylesFor(colorsTheme).timeContainer}>
              {isStarred && (
                <FontAwesome
                  name="star"
                  size={14}
                  color={colorsTheme.textColor}
                  style={{ marginRight: 5 }}
                />
              )}
              <Text style={stylesFor(colorsTheme).time}>{dateString}</Text>
            </View>
          )}

          <Menu name={id.current} ref={menuRef}>
            <MenuTrigger />

            <MenuOptions>
              <MenuItem
                text="Copy to clipboard"
                icon={'copy'}
                onSelect={() => copyToClipboard(text)}
              />
              <MenuItem
                text={`${isStarred ? 'Unstar' : 'Star'} message`}
                icon={isStarred ? 'star-o' : 'star'}
                iconPack={FontAwesome}
                onSelect={() => starMessage(messageId, chatId, userId)}
              />
              <MenuItem
                text="Reply"
                icon="arrow-left-circle"
                onSelect={setReply}
              />
            </MenuOptions>
          </Menu>
        </View>
      </Container>
    </View>
  )
}

const stylesFor = (colorsTheme) =>
  StyleSheet.create({
    wrapperStyle: {
      flexDirection: 'row',
      justifyContent: 'center',
    },
    container: {
      backgroundColor: 'white',
      borderRadius: 6,
      padding: 5,
      marginBottom: 10,
      borderColor: colorsTheme.bubbleBorderColor,
      borderWidth: 1,
    },
    text: {
      fontFamily: 'regular',
      letterSpacing: 0.3,
    },
    menuItemContainer: {
      flexDirection: 'row',
      padding: 5,
    },
    menuText: {
      flex: 1,
      fontFamily: 'regular',
      letterSpacing: 0.3,
      fontSize: 16,
    },
    timeContainer: {
      flexDirection: 'row',
      justifyContent: 'flex-end',
    },
    time: {
      fontFamily: 'regular',
      letterSpacing: 0.3,
      color: colorsTheme.grey,
      fontSize: 12,
    },
    name: {
      fontFamily: 'medium',
      letterSpacing: 0.3,
      color: colorsTheme.bubbleTextColor,
    },
    image: {
      width: 300,
      height: 300,
      marginBottom: 5,
    },
  })

export default Bubble
