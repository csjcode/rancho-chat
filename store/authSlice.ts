import { PayloadAction, createSlice } from "@reduxjs/toolkit";

interface AuthState {
  token: string | null;
  userData: object | null;
  didTryAutoLogin: boolean;
}

const initialState: AuthState = {
  token: null,
  userData: null,
  didTryAutoLogin: false,
};

const authSlice = createSlice({
  name: "auth",
  initialState: {
    token: null,
    userData: null,
    didTryAutoLogin: false,
  } as AuthState,
  reducers: {
    authenticate: (
      state,
      action: PayloadAction<{ token: string; userData: object }>
    ) => {
      const { payload } = action;
      state.token = payload.token;
      state.userData = payload.userData;
      state.didTryAutoLogin = true;
    },
    setDidTryAutoLogin: (state, action) => {
      state.didTryAutoLogin = true;
    },
    logout: (state, action) => {
      state.token = null;
      state.userData = null;
      state.didTryAutoLogin = false;
    },
    updateLoggedInUserData: (
      state,
      action: PayloadAction<{ newData: object }>
    ) => {
      state.userData = { ...state.userData, ...action.payload.newData };
    },
  },
});

export const setDidTryAutoLogin = authSlice.actions.setDidTryAutoLogin;
export const authenticate = authSlice.actions.authenticate;
export const updateLoggedInUserData = authSlice.actions.updateLoggedInUserData;
export const logout = authSlice.actions.logout;
export default authSlice.reducer;
