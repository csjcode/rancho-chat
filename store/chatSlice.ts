import { PayloadAction, createSlice } from "@reduxjs/toolkit";

interface ChatsState {
  chatsData: { [key: string]: any };
}

const initialState: ChatsState = {
  chatsData: {},
};

const chatSlice = createSlice({
  name: "chats",
  initialState,
  reducers: {
    setChatsData: (state, action: PayloadAction<ChatsState>) => {
      state.chatsData = { ...action.payload.chatsData };
    },
  },
});

export const { setChatsData } = chatSlice.actions;
export default chatSlice.reducer;
