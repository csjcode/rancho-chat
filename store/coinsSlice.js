import { createSlice } from '@reduxjs/toolkit'

const coinsSlice = createSlice({
  name: 'coins',
  initialState: {
    storedCoins: {
      coins: ['solana', 'chainlink', 'stepn', 'audius', 'msol'],
    },
  },
  reducers: {
    setStoredCoins: (state, action) => {
      console.log(`action ${JSON.stringify(action)}`)
      const newCoins = action.payload
      state.storedCoins = newCoins
    },
    removeStoredCoin: (state, action) => {
      // {"type":"coins/removeStoredCoin","payload":{"coins":["solana","chainlink","stepn","audius","msol"],"removeCoin":"audius"}}
      console.log(`removeStoredCoin ${JSON.stringify(action)}`)
      const newCoins = action.payload.coins.filter(
        (item) => item !== action.payload.removeCoin,
      )
      // console.log(`action ${JSON.stringify(action)}`)
      // const newCoins = state.storedCoins.coins.pop(action.payload)
      state.storedCoins.coins = newCoins
    },
    addStoredTokenList: (state, action) => {
      console.log(`action ${JSON.stringify(action.payload)}`)
      const newCoins = Array.from(
        new Set([
          ...action.payload.coins,
          ...action.payload.addTokenListStored,
        ]),
      )
      console.log(`newCoins ${newCoins}`)
      state.storedCoins.coins = newCoins
    },
  },
})
export const setStoredCoins = coinsSlice.actions.setStoredCoins
export const removeStoredCoin = coinsSlice.actions.removeStoredCoin
export const addStoredTokenList = coinsSlice.actions.addStoredTokenList
export default coinsSlice.reducer
