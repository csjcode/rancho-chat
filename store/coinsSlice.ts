
import { createSlice } from '@reduxjs/toolkit';

interface StoredCoinsState {
  storedCoins: {
    coins: string[];
  };
}

interface SetStoredCoinsAction {
  payload: StoredCoinsState['storedCoins'];
}

interface RemoveStoredCoinAction {
  payload: {
    coins: string[];
    removeCoin: string;
  };
}

interface AddStoredTokenListAction {
  payload: {
    coins: string[];
    addTokenListStored: string[];
  };
}

const coinsSlice = createSlice({
  name: 'coins',
  initialState: {
    storedCoins: {
      coins: ['solana', 'chainlink', 'stepn', 'audius', 'msol'],
    },
  },
  reducers: {
    setStoredCoins: (state, action: SetStoredCoinsAction) => {
      console.log(`action ${JSON.stringify(action)}`);
      const { payload } = action;
      state.storedCoins = payload;
    },
    removeStoredCoin: (state, action: RemoveStoredCoinAction) => {
      console.log(`removeStoredCoin ${JSON.stringify(action)}`);
      const { payload } = action;
      const { coins, removeCoin } = payload;
      const newCoins = coins.filter((item) => item !== removeCoin);
      state.storedCoins.coins = newCoins;
    },
    addStoredTokenList: (state, action: AddStoredTokenListAction) => {
      console.log(`action ${JSON.stringify(action.payload)}`);
      const { payload } = action;
      const { coins, addTokenListStored } = payload;
      const newCoins = Array.from(
        new Set([...coins, ...addTokenListStored]),
      );
      console.log(`newCoins ${newCoins}`);
      state.storedCoins.coins = newCoins;
    },
  },
});
export const setStoredCoins = coinsSlice.actions.setStoredCoins
export const removeStoredCoin = coinsSlice.actions.removeStoredCoin
export const addStoredTokenList = coinsSlice.actions.addStoredTokenList
export default coinsSlice.reducer
