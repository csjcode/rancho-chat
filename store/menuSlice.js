import { createSlice } from '@reduxjs/toolkit'

const menuSlice = createSlice({
  name: 'menu',
  initialState: {
    storedMenu: {
      map: true,
      tricks: true,
      themeColorLight: false,
    },
  },
  reducers: {
    setStoredMenu: (state, action) => {
      // console.log(`action ${JSON.stringify(action)}`)
      const newMenu = action.payload
      state.storedMenu = newMenu
      // console.log(state.storedMenu)
    },
  },
})
export const setStoredMenu = menuSlice.actions.setStoredMenu
export default menuSlice.reducer
