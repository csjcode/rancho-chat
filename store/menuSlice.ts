import { PayloadAction, createSlice } from "@reduxjs/toolkit";

interface StoredMenu {
  map: boolean;
  tricks: boolean;
  themeColorsName: string;
}

interface MenuState {
  storedMenu: StoredMenu;
}

const initialState: MenuState = {
  storedMenu: {
    map: true,
    tricks: true,
    themeColorsName: "dark",
  },
};

const menuSlice = createSlice({
  name: "menu",
  initialState,
  reducers: {
    setStoredMenu: (state, action: PayloadAction<StoredMenu>) => {
      // console.log(`action ${JSON.stringify(action)}`)
      const newMenu = action.payload;
      state.storedMenu = newMenu;
    },
  },
});
export const setStoredMenu = menuSlice.actions.setStoredMenu;
export default menuSlice.reducer;

/*


import { createSlice } from '@reduxjs/toolkit'

const menuSlice = createSlice({
  name: 'menu',
  initialState: {
    storedMenu: {
      map: true,
      tricks: true,
      themeColorsName: 'dark',
    },
  },
  reducers: {
    setStoredMenu: (state, action) => {
      // console.log(`action ${JSON.stringify(action)}`)
      const newMenu = action.payload
      state.storedMenu = newMenu
    },
  },
})
export const setStoredMenu = menuSlice.actions.setStoredMenu
export default menuSlice.reducer

*/
