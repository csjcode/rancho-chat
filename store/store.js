import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit'
import { persistReducer, persistStore } from 'redux-persist'

import AsyncStorage from '@react-native-async-storage/async-storage'
import authSlice from './authSlice'
import chatSlice from './chatSlice'
import coinsSlice from './coinsSlice'
import menuSlice from './menuSlice'
import messagesSlice from './messagesSlice'
import notesSlice from './notesSlice'
import userSlice from './userSlice'

const persistNotesConfig = {
  key: 'notes',
  storage: AsyncStorage,
}

const persistedNotes = persistReducer(persistNotesConfig, notesSlice)

export const store = configureStore({
  reducer: {
    auth: authSlice,
    users: userSlice,
    chats: chatSlice,
    messages: messagesSlice,
    menu: menuSlice,
    coins: coinsSlice,
    notes: persistedNotes,
  },
  middleware: getDefaultMiddleware({
    serializableCheck: false,
  }),
})
export const persistor = persistStore(store)
