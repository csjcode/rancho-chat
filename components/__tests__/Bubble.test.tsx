import { MenuProvider } from "react-native-popup-menu";
import { Provider } from "react-redux";
import ReactTestRenderer from "react-test-renderer";
import { store } from "../../store/store";
import Bubble from "../Bubble";
import * as Clipboard from "expo-clipboard";

jest.mock("expo-clipboard", () => ({
  setStringAsync: jest.fn(),
}));

describe("Bubble snapshot is ok.", () => {
  xit("Bubble snapshot test", async () => {
    const tree = ReactTestRenderer.create(
      <Provider store={store}>
        <MenuProvider>
          <Bubble
            text="Hello, world!"
            type=""
            messageId="123"
            chatId="456"
            userId="789"
            date={new Date()}
            setReply={() => {}}
            replyingTo={{ sentBy: "abc", messageId: "def" }}
            name="John Doe"
            imageUrl="https://example.com/image.png"
          />
        </MenuProvider>
      </Provider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
