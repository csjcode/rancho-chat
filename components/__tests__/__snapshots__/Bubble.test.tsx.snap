// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bubble snapshot is ok. Bubble snapshot test 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "flex-end",
        }
      }
    >
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#1b4e3f",
            "borderColor": "#7f8c8d",
            "borderRadius": 6,
            "borderWidth": 1,
            "marginBottom": 10,
            "maxWidth": "90%",
            "padding": 5,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#dfdfdf",
              "fontFamily": "medium",
              "letterSpacing": 0.3,
            }
          }
        >
          John Doe
        </Text>
        <Image
          source={
            Object {
              "uri": "https://example.com/image.png",
            }
          }
          style={
            Object {
              "height": 300,
              "marginBottom": 5,
              "width": 300,
            }
          }
        />
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "flex-end",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#7f8c8d",
                "fontFamily": "regular",
                "fontSize": 12,
                "letterSpacing": 0.3,
              }
            }
          >
            2:13 pm
          </Text>
        </View>
        <View>
          <View
            collapsable={false}
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
            >
              <View
                ctx={
                  Object {
                    "menuActions": Object {
                      "_getOpenedMenu": [Function],
                      "_notify": [Function],
                      "closeMenu": [Function],
                      "isMenuOpen": [Function],
                      "openMenu": [Function],
                      "toggleMenu": [Function],
                    },
                    "menuRegistry": Object {
                      "getAll": [Function],
                      "getMenu": [Function],
                      "setOptionsCustomStyles": [Function],
                      "subscribe": [Function],
                      "unsubscribe": [Function],
                      "updateLayoutInfo": [Function],
                    },
                  }
                }
                style={
                  Array [
                    undefined,
                    undefined,
                  ]
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <RCTSafeAreaView
    pointerEvents="box-none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      collapsable={false}
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </RCTSafeAreaView>
</View>
`;
